### PR TITLE
Adds DeltaSharingSourceOffset

### PR DESCRIPTION
**Pull request by linzhou-db**
Adds DeltaSharingSourceOffset, similar to https://github.com/delta-io/delta/blob/master/core/src/main/scala/org/apache/spark/sql/delta/sources/DeltaSourceOffset.scala

### CODE REVIEW

File: spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSourceOffset.scala [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+import org.json4s._
+import org.json4s.jackson.JsonMethods.parse
+
+import io.delta.sharing.spark.util.JsonUtils
+
+/**
+ * Tracks how far we processed in when reading changes from the [[Delta Sharing Server]].
+ *
+ * Note this class retains the naming of `Reservoir` to maintain compatibility
+ * with serialized offsets from the beta period.
```
Comment (chakankardb) [inline spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSourceOffset.scala:31]: Is this applicable to us?

File: spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSourceOffset.scala [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+ *
+ * @param sourceVersion     The version of serialization that this offset is encoded with.
+ * @param reservoirId       The id of the table we are reading from. Used to detect
+ *                          misconfiguration when restarting a query.
+ * @param reservoirVersion  The version of the table that we are currently processing.
```
Comment (chakankardb) [inline spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSourceOffset.scala:36]: Do we need to use the reservoir prefixes? Can we call them tableId/tableVersion?

File: spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSourceOffset.scala (hunk 1 above)
Comment (linzhou-db) [reply spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSourceOffset.scala]: good catch, I actually asked @zsxwing about this and it doesn't apply to us, and forgot to update it here.
Changed to tableId and tableVersion

File: spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSourceOffset.scala (hunk 2 above)
Comment (linzhou-db) [reply spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSourceOffset.scala]: updated.

File: spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSourceOffset.scala [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+      throw new IllegalStateException(
+        s"Found invalid offsets: 'reservoirVersion' moved back. " +
+          s"Previous: $previousOffset, Current: $currentOffset")
+    }
+    if (previousOffset.reservoirVersion == currentOffset.reservoirVersion &&
+      previousOffset.index > currentOffset.index) {
+      throw new IllegalStateException(
+        s"Found invalid offsets. 'index' moved back. " +
+          s"Previous: $previousOffset, Current: $currentOffset")
+    }
```
[the remark this answers is not in the document]
Comment (linzhou-db) [reply spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSourceOffset.scala:138]: changed to `compare`, which will miss the detailed message on "what moved back", but I feel it's ok.

File: spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSourceOffsetSuite.scala [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+      """.stripMargin
+    val e = intercept[IllegalStateException] {
+      DeltaSharingSourceOffset(UUID.randomUUID().toString, SerializedOffset(json))
+    }
+    for (msg <- Seq("Cannot find", "sourceVersion")) {
+      assert(e.getMessage.contains(msg))
+    }
+  }
+
+  test("DeltaSharingSourceOffset - unmatched reservoir id") {
```
Comment (chakankardb) [inline spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSourceOffsetSuite.scala:89]: reservoir?

File: spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSourceOffsetSuite.scala [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+        tableVersion = 4,
+        index = 10,
+        isStartingVersion = false),
+      currentOffset = DeltaSharingSourceOffset(
+        sourceVersion = 1,
+        tableId = "foo",
+        tableVersion = 5,
+        index = 1,
+        isStartingVersion = false)
+    )
```
Comment (chakankardb) [inline spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSourceOffsetSuite.scala:136]: consider adding the following:
- we only increase the index
- isStartingVersion is true for both (we can perhaps do a grid test for success where all combinations can be tested for isStartingVersion being false and true).

File: spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSourceOffsetSuite.scala (hunk 4 above)
Comment (linzhou-db) [reply spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSourceOffsetSuite.scala]: removed.

File: spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSourceOffsetSuite.scala (hunk 5 above)
Comment (linzhou-db) [reply spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSourceOffsetSuite.scala]: updated.

File: spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSourceOffsetSuite.scala [hunk 6]
```diff
@@ -106,82 +106,53 @@ class DeltaSharingSourceOffsetSuite extends QueryTest
...
+    testValidateOffset(4, 10, false, 5, 1, false, None)
+    testValidateOffset(4, 10, true, 4, 10, true, None)
+    testValidateOffset(4, 10, true, 4, 11, true, None)
+    testValidateOffset(4, 10, true, 5, 1, true, None)
+
+    // errors on backward moving offset
+    testValidateOffset(4, 10, false, 4, 9, false, Some("Found invalid offsets. Previous:"))
+    testValidateOffset(4, 10, false, 3, 11, false, Some("Found invalid offsets. Previous:"))
+    testValidateOffset(4, 10, true, 4, 9, true, Some("Found invalid offsets. Previous:"))
+    testValidateOffset(4, 10, true, 3, 11, true, Some("Found invalid offsets. Previous:"))
```
Comment (chakankardb) [inline spark/src/test/scala/io/delta/sharing/spark/DeltaSharingSourceOffsetSuite.scala:151]: nice tests!

**Review by chakankardb — APPROVED** [review]

**Pull request merged**